File: jump.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach
/*
 * This approach is similar to bfs where we process the numbers level wise, and find the max index we can reach from each level.
 * Once a level is processed we will increment the count.
 */

class Solution {
    public int jump(int[] nums) {
        int n = nums.length;
        int left = 0, right = 0, farthest = 0;
        int count = 0;

        while(right < n){
            if(farthest == n-1)
                break;
                
            for(int i=left; i<=right; i++){
                farthest = Math.max(farthest, nums[i] + i);
            }
            left = right+1;
            right = farthest;
            count++;
        }

        return count;
    }
}

File: Jump.py
# Time complexity - O(n)
#  Space complexity - O(1)
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        n=len(nums)
        if(n==1):
            return True
        target = n-1
        for i in range(n-2, -1, -1):
            if i+nums[i]>=target:
                target=i
        
        if target == 0:
            return True
        else:
            return False